### PR TITLE
Device ordinal doc

### PR DESCRIPTION
**Pull request by trivialfis**
- Rewrite GPU demos. notebook is converted to script to avoid committing additional png plots.
- Add GPU demos into the sphinx gallery.
- Add RMM demos into the sphinx gallery.
- Test for firing threads with different device ordinals.

### CODE REVIEW

**Comment by trivialfis** [comment]
There are other mentions of `gpu_hist` scattered in doc strings. I will do a complete scan later.

File: demo/gpu_acceleration/tree_shap.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    "tree_method": "hist",
+    "device": "cuda",
+}
+
+# GPU accelerated training
+dtrain = xgb.DMatrix(X, label=y, feature_names=data.feature_names)
+model = xgb.train(param, dtrain, num_round)
+
+# Compute shap values using GPU with xgboost
+model.set_param({"device": "cuda"})
```
Comment (hcho3) [inline demo/gpu_acceleration/tree_shap.py:33]: Is this line necessary?

File: src/data/data.cc [hunk 2]
```diff
@@ -724,11 +724,15 @@ void MetaInfo::SynchronizeNumberOfColumns() {
...
-  CHECK(v.DeviceIdx() == Context::kCpuId || device == Context::kCpuId || v.DeviceIdx() == device)
-      << "Data is resided on a different device than `gpu_id`. "
-      << "Device that data is on: " << v.DeviceIdx() << ", "
-      << "`gpu_id` for XGBoost: " << device;
+  bool valid =
+      v.DeviceIdx() == Context::kCpuId || device == Context::kCpuId || v.DeviceIdx() == device;
+  if (!valid) {
+    LOG(FATAL) << "Invalid data ordinal, data is resided on a different device than the ordinal of "
+                  "the booster. The ordinal of the data is: "
+               << v.DeviceIdx() << ", the ordinal for XGBoost is: " << device;
```
Comment (hcho3) [inline src/data/data.cc:732]:
```suggestion
    LOG(FATAL) << "Invalid device ordinal, data is associated with a different device ordinal than "
                  "the Booster object. The device ordinal of the data is: "
               << v.DeviceIdx() << "; the device ordinal of the Booster object is: " << device;
```

Users may not understand what `ordinal` is. Let's use `device ordinal` to make it clear.

File: tests/python-gpu/test_from_cupy.py [hunk 3]
```diff
@@ -234,7 +234,7 @@ def test_specified_device(self):
         cp.cuda.runtime.setDevice(0)
         dtrain = dmatrix_from_cupy(np.float32, xgb.QuantileDMatrix, np.nan)
         with pytest.raises(
-            xgb.core.XGBoostError, match="Data is resided on a different device"
+            xgb.core.XGBoostError, match="Invalid data ordinal"
```
Comment (hcho3) [inline tests/python-gpu/test_from_cupy.py:237]:
```suggestion
            xgb.core.XGBoostError, match="Invalid device ordinal"
```

File: tests/python-gpu/test_gpu_with_sklearn.py [hunk 4]
```diff
@@ -161,3 +163,46 @@ def test_ranking_qid_df():
...
+        reg = xgb.XGBRegressor(device=f"cuda:{ordinal}", tree_method="hist")
+
+        if correct_ordinal:
+            reg.fit(
+                X, y, sample_weight=w, eval_set=[(X, y)], sample_weight_eval_set=[w]
+            )
+            assert tm.non_increasing(reg.evals_result()["validation_0"]["rmse"])
+            return
+
+        with pytest.raises(ValueError, match="Invalid data ordinal"):
```
Comment (hcho3) [inline tests/python-gpu/test_gpu_with_sklearn.py:191]:
```suggestion
        with pytest.raises(ValueError, match="Invalid device ordinal"):
```

File: doc/tutorials/monotonic.rst [hunk 5]
```diff
@@ -83,13 +83,12 @@ Some other examples:
...
 
-**Note for the 'hist' tree construction algorithm**.
-If ``tree_method`` is set to either ``hist``, ``approx`` or ``gpu_hist``, enabling
-monotonic constraints may produce unnecessarily shallow trees. This is because the
-``hist`` method reduces the number of candidate splits to be considered at each
-split. Monotonic constraints may wipe out all available split candidates, in which case no
-split is made. To reduce the effect, you may want to increase the ``max_bin`` parameter to
-consider more split candidates.
+**Note for the 'hist' tree construction algorithm**.  If ``tree_method`` is set to either
+``hist``, ``approx``, enabling monotonic constraints may produce unnecessarily shallow
```
Comment (hcho3) [inline doc/tutorials/monotonic.rst:87]:
```suggestion
``hist`` or ``approx``, enabling monotonic constraints may produce unnecessarily shallow
```

File: demo/rmm_plugin/README.rst [hunk 6]
```diff
@@ -0,0 +1,51 @@
+Using XGBoost with RAPIDS Memory Manager (RMM) plugin (EXPERIMENTAL)
+====================================================================
+
+[RAPIDS Memory Manager `RMM <https://github.com/rapidsai/rmm>`__ library provides a
```
Comment (hcho3) [inline demo/rmm_plugin/README.rst:4]:
```suggestion
`RAPIDS Memory Manager (RMM) <https://github.com/rapidsai/rmm>`_ library provides a
```

File: demo/rmm_plugin/README.rst [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+collection of efficient memory allocators for NVIDIA GPUs. It is now possible to use
+XGBoost with memory allocators provided by RMM, by enabling the RMM integration plugin.
+
+The demos in this directory highlights one RMM allocator in particular: **the pool
+sub-allocator**.  This allocator addresses the slow speed of `cudaMalloc()` by allocating
+a large chunk of memory upfront. Subsequent allocations will draw from the pool of already
+allocated memory and thus avoid the overhead of calling `cudaMalloc()` directly. See `this
+GTC talk slides
+<https://on-demand.gputechconf.com/gtc/2015/presentation/S5530-Stephen-Jones.pdf>`_ for
+more details.
```
Comment (hcho3) [inline demo/rmm_plugin/README.rst:14]:
```suggestion
The demos in this directory highlights one RMM allocator in particular: **the pool
sub-allocator**.  This allocator addresses the slow speed of ``cudaMalloc()`` by allocating
a large chunk of memory upfront. Subsequent allocations will draw from the pool of already
allocated memory and thus avoid the overhead of calling ``cudaMalloc()`` directly. See `this
GTC talk slides
<https://on-demand.gputechconf.com/gtc/2015/presentation/S5530-Stephen-Jones.pdf>`_ for
more details.
```

File: demo/rmm_plugin/README.rst [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+Before running the demos, ensure that XGBoost is compiled with the RMM plugin enabled. To do this,
+run CMake with option `-DPLUGIN_RMM=ON` (`-DUSE_CUDA=ON` also required):
```
Comment (hcho3) [inline demo/rmm_plugin/README.rst:17]:
```suggestion
run CMake with option ``-DPLUGIN_RMM=ON`` (``-DUSE_CUDA=ON`` also required):
```

**Review by hcho3 — CHANGES_REQUESTED** [review]

File: demo/gpu_acceleration/tree_shap.py (hunk 1 above)
Comment (trivialfis) [reply demo/gpu_acceleration/tree_shap.py:33]: Only for demo

File: demo/rmm_plugin/README.rst (hunk 6 above)
Comment (trivialfis) [reply demo/rmm_plugin/README.rst:4]: This is actually correct, we use double underscore to prevent duplicated names.

Comment (hcho3) [reply demo/rmm_plugin/README.rst:4]: Okay, you should still remove the square bracket in the beginning

Comment (trivialfis) [reply demo/rmm_plugin/README.rst:4]: Good catch!

**Review by hcho3 — APPROVED** [review]

**Pull request merged**